@catch, retry, partition, asyncmap and pmap

samoconnor opened the pull request:
Third step in [Simplifying and generalising pmap (as per #14843)](https://github.com/JuliaLang/julia/issues/14843).

See also [#15058 StreamMapIterator (was asyncmap)](https://github.com/JuliaLang/julia/pull/15058) and [#15073  WorkerPool and remote](https://github.com/JuliaLang/julia/pull/15073)
### @catch

`@catch` takes a function and returns a lambda that catches any exceptions thrown by the function and returns the exception object.

``` julia
@test map(typeof, map(@catch(i->[1,2,3][i]), 1:6)) ==
      [Int, Int, Int, BoundsError, BoundsError, BoundsError]
```
### retry

`retry` takes a function and returns a lambda that retries the function 3 times if an error occurs

``` julia
#pmap(f, v; err_retry=true)
pmap(retry(f), v)

retry(http_get, e->e.status == "503")(url)

retry(read, UVError)(io)

retry(read, UVError, n=10)(io) # retry 10 times...
```
### partition(collection, n)

`partition(collection, n)` returns an `n` elements at a time iterator over collection.

``` julia
julia> collect(partition([1,2,3,4,5], 2))
3-element Array{Array{Int64,1},1}:
 [1,2]
 [3,4]
 [5]
```
### batchsplit(c; min_batch_count=1, max_batch_size=100)

Split a collection into at least `min_batch_count` batches. 

Equivalent to `partition(c, max_batch_size)` when `length(c) >> max_batch_size`.
### asyncgenerate and asyncmap

Generate or map using parallel tasks.

``` julia
asyncgenerate(f, c) = StreamGenerator(f, c)
asyncmap(f, c) = collect(asyncgenerate(f, c))
```
### pgenerate and pmap

Generate or map using parallel workers and tasks.

``` julia
function pgenerate(f, c)
    batches = batchsplit(c, min_batch_count = nworkers() * 3)
    return flatten(asyncgenerate(remote(b -> asyncmap(f, b)), batches))
end

pmap(f, c) = collect(pgenerate(f, c))
```

samoconnor committed:
@catch, retry, partition, asyncmap and refactored pmap (Seep #15409 and #14843)

Rename *MapIterator to *Generator

tkelman commented on test/error.jl:
```diff
@@ -0,0 +1,72 @@
+# This file is a part of Julia. License is MIT: http://julialang.org/license
```
this needs to be added to `test/choosetests.jl` to actually run

samoconnor replied:
true that 🙄

samoconnor commented:
Note: this commit adds a small default delay to `retry` with exponential back-off and some random jitter: https://github.com/samoconnor/julia/commit/88ea1a13e85cfb0ce1bc0bce72fc55a4a701b2a1

I'm happy to back this out if it's seen as adding too much at once.

However, it seems to me that it is probably a good thing to have a small delay by default when retrying something that caused an exception (or at the very least to `yield()`). `retry` is only useful for non-deterministic calls. It seems reasonable to assume that the primary source of non-determinism is likely to be an external or remote file or process or network or database. In all these cases exponential back-off is usually a good thing for overall system performance.

samoconnor commented:
rebased to ec6f886a0d938046ae3adb47cf2b9cde6a5c1621

samoconnor commented:
tests now ok.
default constructor for WorkerPool added per https://github.com/JuliaLang/julia/pull/15073#discussion_r55512302

tkelman commented:
`collect(n, collection)` returning an iterator doesn't mesh very well with all other methods of `collect`. That should possibly go by a different name.

samoconnor commented:
@JeffBezanson any suggestions on what this thing should be called if not `collect` : https://github.com/samoconnor/julia/commit/0483198785638f20d077440632ddd74f753493a9#diff-83c91f45204e96e58535f5491cca8cd8R347

`eachchunk` ?
`eachn` ?

amitmurthy commented on base/workerpool.jl:
```diff
@@ -3,6 +3,9 @@
     channel::RemoteChannel{Channel{Int}}
+
+    # Create a shared queue of available workers...
+    WorkerPool() = new(RemoteChannel(()->Channel{Int}(128)))
```
This should be changed to `RemoteChannel(()->Channel(typemax(Int)))`, else ` WorkerPool(workers::Vector{Int})` will block when trying to add more than 128 workers.

samoconnor replied:
will do

amitmurthy commented on base/error.jl:
```diff
@@ -49,3 +49,47 @@ macro assert(ex, msgs...)
+    end
+end
+
+retry(f::Function, condition::Type, n::Int=3) = retry(f, e->isa(e, condition), n)
```
`retry(f::Function, condition::Type, n::Int=3; kwargs) = retry(f, e->isa(e, condition), n; kwargs...)` will allow supporting `max_delay` in both signatures.

`n::Int=3` should be `n=NUM_DEFAULT_RETRY` so that the value is defined in only one place.

samoconnor replied:
I've used the `kw...` pattern to handle single-point-of-definition for both `n` and `max_delay`.

amitmurthy commented:
`split`? On strings it returns the split strings. But since an iterator is being returned here, maybe just `chunk`?

samoconnor commented:
There is some precedence for things that return an iterator having an `each` prefix.
 - [`eachindex`](http://docs.julialang.org/en/latest/stdlib/arrays/?highlight=each#Base.eachindex)
 - [`eachline`](http://docs.julialang.org/en/latest/stdlib/io-network/?highlight=each#Base.eachline)
 - [`eachmatch`](http://docs.julialang.org/en/latest/stdlib/strings/?highlight=each#Base.eachmatch)
 
Then there are a bunch of things that have no hint of iterator-ness in their names:
`enumerate`, `zip`, `filter`, `cycle`, `rest`, `countfrom`, `take`, `drop`, `repeated`, `keys`, `values`, `graphemes`

`split` would be consistent with `enumerate`, `cycle`, `filter` and `zip`
`chunks` would be consistent with `graphemes`, `keys` and `values`...

[`batch`](http://sadi.googlecode.com/svn/trunk/sadi.common/src/main/java/ca/elmonline/util/BatchIterator.java) ?, [`partition`](http://docs.guava-libraries.googlecode.com/git/javadoc/com/google/common/collect/Lists.html#partition%28java.util.List,%20int%29) ?,  `slice` ?, `divide` ?, `subdivide` ?, `segment` ?, `eachblock` ?, `blocks` ?

samoconnor commented:
On reflection, @amitmurthy's suggestion of `split(c, n)` is good. Changed.

It would have been very handy for this https://github.com/samoconnor/AWSCore.jl/commit/9330a7b5131dde536ff6645f25822dd552c2c398#diff-c33c63b7ba6d9fce289488d1cf3dfcaeR64 just yesterday.

nalimilan commented:
Isn't `collect` what Iterators.jl calls `partition`? There's precedent of moving functions from there into Base.

samoconnor commented:
> Isn't collect what Iterators.jl calls partition?

Thanks for pointing that out @nalimilan, its good to know about Iterators.jl.
However, while `partition` is similar, it isn't equivelant to ~~`collect(n, c)`~~ `split(c, n)`
Partition always produces `Tuple`s and seems to ignore the tail of the collection...

```julia
julia> collect(partition("Hello World!", 5))
2-element Array{Tuple{Char,Char,Char,Char,Char},1}:
 ('H','e','l','l','o')
 (' ','W','o','r','l')

julia> collect(split("Hello World!", 5))
3-element Array{ASCIIString,1}:
 "Hello"
 " Worl"
 "d!"

julia> collect(partition([1,2,3,4,5], 3))
1-element Array{Tuple{Int64,Int64,Int64},1}:
 (1,2,3)

julia> collect(split([1,2,3,4,5], 3))
2-element Array{Array{Int64,1},1}:
 [1,2,3]
 [4,5]
```

amitmurthy commented:
Sorry for the delay reviewing this.

@samoconnor , I am thinking it will be easier to just have the complete new `pmap` in this PR with minimal exports. Separate smaller PRs is usually better, but here it appears that we are losing out on the bigger context of changes being done to generalize `pmap` and the need to add new exports if any.

Avoiding new exports makes the PR less controversial since we are not adding something that needs to be supported over the long term.

I'll push a fix separately for the `WorkerPool` and mapiterator corrections (since they do not really belong here).

amitmurthy commented:
> I'll push a fix separately for the WorkerPool and mapiterator corrections (since they do not really belong here).

Unless you would rather submit a PR for these corrections. I should have asked.

nalimilan commented:
@samoconnor Looks like you should add a `partial::Bool=false` argument to `partition` then. As regards tuples vs. arrays, we cannot duplicate all functions for which this choice exists, so we have to find a solution.

samoconnor commented:
@amitmurthy, please go ahead and make the `WorkerPool` and `mapiterator` corrections on `master`. Thanks. I'll rebase this PR when that's done.

I'll also add `pmap` to this PR as you suggest.

My current intention is for `pmap` to do something like this by default:

```julia
asyncmap(f, c...) = collect(StreamMapIterator(f, c...))

pmap(f, c...) = vcat(asyncmap(remote(chunk->asyncmap(a->f(a...), chunk), split(zip(c), 100))
```

i.e. run batches of 100 (or whatever a good default is) async maps on each worker.
For IO bound code, the async-ness on each worker is is a big win.
For compute-bound code, the overhead is probably minimal. I assume that if the mapped function never `yields` then the async map is almost the same as normal map.
Of course this can be made tuneable with kw args.

samoconnor commented:
@amitmurthy

I've just committed a WIP pmap() implementation.
It's not really ready for review, needs doc, tests etc, but it might be useful for you to see the direction I'm going in.

I've also included `flatten` from https://github.com/JuliaLang/julia/pull/14805

samoconnor commented:
- Commits cleaned up a bit and rebased/squashed.
 - Updated main PR description above https://github.com/JuliaLang/julia/pull/15409#issue-139414966

samoconnor commented:
CI all passing exepct MS i686 ? is that just AppVeyor being odd?

tkelman commented on base/generator.jl:
```diff
@@ -20,6 +20,12 @@ function next(g::Generator, s)
+Apply `f` to each element of `c`.
+For multiple collection arguments, apply f elementwise.
+"""
 generate(f, c...) = Generator(f, c...)
```
please don't export this. It's a very small wrapper and unfortunately close in syntax to `@generated`

nalimilan replied:
Why not call `Generator` directly?

samoconnor replied:
I guess, because it feels to me like it should not be necessary to think about types to use a Julia API.

It seems like there is a greater burden of understanding if I have to call a type constructor. If I'm making something explicitly this type, what else do I have to understand about this type?

If I just call a function, the type will be whatever clever type the method dispatch system figures out is best for the particular situation. Maybe `generate(f, c...)` will return a different type for some values of `f` and `c...`.

Note, that this is not actually called anywhere. It is included mostly for completeness to establish the pattern followed by `asyncgenerate` and `pgenerate`.

Also, there are precedents for having a `verb` function to create an iterator: `zip`, `filter`, `enumerate` etc.

nalimilan replied:
I see your point. Clearly, we don't know what policy to follow as regards the naming of iterators and functions that create them...

Anyway, the docs should explicitly say "Return an iterator applying `f` to each element of `c`.".

samoconnor replied:
Isn't that what `-> iterator` implies?

nalimilan replied:
Yes, but it's still confusing to say "apply" when all the function does is create a generator. Better be explicit.

samoconnor replied:
fair enough. done

tkelman replied:
This is just #10452 - though "not having to think about types" sounds like you're using the wrong language. You have to think about types all the time with Julia, and unless there are actual examples of a lowercase function API returning different types for different inputs, exporting a new function that's just a trivial wrapper around an existing type is not necessary.

nalimilan replied:
@tkelman I agree with you, but in https://github.com/JuliaLang/julia/issues/10162 no clear decision was taken. I would have preferred to have only upper case constructors for all this family of iterators which are always of the same type. Maybe restart the discussion there?

samoconnor replied:
Why wouldn't you want to allow `generate` to return any one of `FallbackGenerator`, `GeneratorSpecialisedForVectors`, `GeneratorOptimisedForInputOfKnownLength`, `GeneratorOptimisedForSingleInputCollection`, etc, etc. ?

```julia
julia> typeof(zip([1]))
Base.Zip1{Array{Int64,1}}

julia> typeof(zip([1],[2]))
Base.Zip2{Array{Int64,1},Array{Int64,1}}

julia> typeof(zip([1],[2],[3]))
Zip{Array{Int64,1},Base.Zip2{Array{Int64,1},Array{Int64,1}}}
```

tkelman replied:
If we had any of the latter, I might agree with you. At the moment we don't, so it's redundant to start creating and using API's that don't need to exist yet.

tkelman commented on base/error.jl:
```diff
@@ -49,3 +49,49 @@ macro assert(ex, msgs...)
+
+
+"""
+    retry(f, [condition]; n=3; max_delay=10) -> function
```
Function is capitalized

tkelman commented on base/iterator.jl:
```diff
@@ -123,6 +123,25 @@ done(i::Rest, st) = done(i.itr, st)
+Returns `head`: the first `n` elements of `c`;
+and `tail`: an iterator over the remaining elements.
+"""
+function head_and_tail(c, n)
```
not a good name to export either

tkelman commented on base/iterator.jl:
````diff
@@ -395,3 +414,56 @@ end
+```
+"""
+function split{T}(c::T, n::Int)
+    @assert n > 0
````
ArgumentError, not assert

nalimilan commented:
AppVeyor failure is the same as https://github.com/JuliaLang/julia/pull/15587#issuecomment-200889915.

nalimilan commented:
Regarding `split`: that function already exists for strings, where it splits the string at each occurrence of a character. So the new meaning clearly conflicts with the existing one, in particular given that strings are collections.

I also wonder whether we should find a policy with regard to functions returning iterators. It would be nice to have some consistency, and not have some functions return arrays and other iterators. Maybe use an upper case initial for iterators, since they are actually constructors? This PR currently calls the type `SplitIterator`, which is quite explicit, but I don't think this convention is used elsewhere (is it?). Cf. https://github.com/JuliaLang/Iterators.jl/issues/19.

samoconnor commented:
@nalimilan, re types ending in `Iterator`, there are a few...
```julia
julia> collect(filter(n->ismatch(r"Iterator$",string(n)), names(Base,true)))
8-element Array{Symbol,1}:
 :AbstractProdIterator
 :AbstractZipIterator
 :AsyncMapIterator
 :KeyIterator
 :RegexMatchIterator
 :SplitIterator
 :StreamMapIterator
 :ValueIterator
```

nalimilan commented:
@samoconnor OK, but in the series of simple iterators which have recently been imported into Base (https://github.com/JuliaLang/julia/issues/10162), the convention is different. And I must say your `split` would naturally go with that family.

As for the API, I'm not sure how we could conciliate it with the existing string method. Maybe with two alternative keyword arguments, like `length` and `at`?

samoconnor commented:
@nalimilan re `split` I have included doc for `split(c, n)` in both [strings.rst](https://github.com/JuliaLang/julia/pull/15409/files#diff-c4c0a716d690a268720b61c8706bc0f7R270) and [collections.rst](https://github.com/JuliaLang/julia/pull/15409/files#diff-badd7d54b8cfdba53f7e1064089334adR75) and included a string example doctest to try to make it clear that it works nicely for strings too.

(The implementation has [special handling for strings and vectors](https://github.com/JuliaLang/julia/pull/15409/files#diff-83c91f45204e96e58535f5491cca8cd8R456))

The two split methods for strings co-exist reasonably well:

```julia
julia> split("Hello, World!", ",")
2-element Array{SubString{ASCIIString},1}:
 "Hello"
 " World!"

julia> collect(split("Hello, World!", 4))
4-element Array{ASCIIString,1}:
 "Hell"
 "o, W"
 "orld"
 "!"
```

I would argue that if anything, the existing `split` should return an iterator too.

nalimilan commented:
The issue is that `split` now has two meanings, one of which only applies to strings. If we wanted to generalize that meaning to all collections (which could be useful), it wouldn't be possible because of the ambiguity of `split([1, 2, 3], 2)`. I think the rule in Julia is that a function should only have a single definition.

Also, it doesn't actually work for strings:
```julia
julia> collect(SplitIterator("café crème", 2))
ERROR: UnicodeError: invalid character index
 in getindex(::UTF8String, ::UnitRange{Int64}) at ./unicode/utf8.jl:117
 [inlined code] from ./none:4
 in collect(::Type{UTF8String}, ::SplitIterator{UTF8String}) at ./array.jl:218
 in collect(::SplitIterator{UTF8String}) at ./array.jl:230
 in eval(::Module, ::Any) at ./boot.jl:243
```
But this is a separate implementation issue.

samoconnor commented:
The `UF8String` thing is unfortunate. I hope that will be resolved separately by Stefan's ongoing string refactoring work.

That aside, I'm not fixated on using the name `split`.

There are other suggestions here: https://github.com/JuliaLang/julia/pull/15409#issuecomment-195597749
And you own suggestion (@nalimilan) of `partition`.

I originally chose `split` because @amitmurthy suggested it, and it's him who will most likely merge this PR.

I'm not in favour of spending more time discussing the name in this PR.
@amitmurthy: considering @nalimilan's objections to `split` could you please make the decision on what to call this and I'll update the code accordingly.

nalimilan commented:
> The UF8String thing is unfortunate. I hope that will be resolved separately by Stefan's ongoing string refactoring work.

It won't. Not all indices are valid for strings, period. If we wanted the new `split` to work with strings, it would have to work on code points, which would require a special method. But it's not out of scope for this PR.

I'm not bikeshedding on the name here, but on the use of an existing name for a different operation. I'm still not sure why you cannot use `partition` from Iterators.jl (after adding a `partial` argument): is the fact that it returns a tuple really an issue here?

amitmurthy commented:
@samoconnor , thanks for the `pmap` rewrite and its simplification. The bikeshedding is par for the course in Julia (simply because folks care deeply) and will go through many rounds.

Which is why I suggest we only export names required for the `pmap` rewrite in this PR. Which would be `@catch`, `retry`, `default_worker_pool` and `WorkerPool`? While decisions on which of the other functionality developed can be discussed and exported in other PRs. This will help restrict discussions in this PR to `pmap` only.

amitmurthy commented on base/pmap.jl:
```diff
@@ -0,0 +1,97 @@
+pmap(p::WorkerPool, f, c...) = collect(pgenerate(p, f, c...))
+pmap(f, c...) = pmap(p, f, c...)
+
+function pmap(f, c...; err_retry=nothing, err_stop=nothing, pids=nothing)
```
Currently the defaults are `error_retry=true, error_stop=false`, should keep them the same to avoid unnecessarily breaking user code.

amitmurthy commented on base/pmap.jl:
```diff
@@ -0,0 +1,97 @@
+For multiple collection arguments, apply f elementwise.
+"""
+pmap(p::WorkerPool, f, c...) = collect(pgenerate(p, f, c...))
+pmap(f, c...) = pmap(p, f, c...)
```
Should be `pmap(default_worker_pool(), f, c...)` ?

amitmurthy replied:
How do we handle the case where we want `pmap` to only parallelize over tasks even when `nprocs() > 1` ?

samoconnor replied:
As it stands you could do [`asyncmap(f, c)`](https://github.com/JuliaLang/julia/pull/15409/files/3568a1149486b5b4d3a9ddf31fe98bcbadbd5305#diff-b4c399b423f02e7569b7de53c31795a0R170) or `pmap(WorkerPool(), f, c)` to avoid any remote calls. I guess adding a `tasks_only=true` kw arg would be simple enough. My personal preference would be to just call `asyncmap` in that case.

samoconnor commented:
@amitmurthy ok, I'll change the PR to limit the exports as you suggest, but I still need a different name for `split` because that is already an exported name.

amitmurthy commented:
`partition` or `batch` for now?

amitmurthy commented on base/multi.jl:
```diff
@@ -1514,101 +1514,17 @@ macro everywhere(ex)
 end
 
+#FIXME delete?
 function pmap_static(f, lsts...)
```
This was added in https://github.com/JuliaLang/julia/commit/b325999e11a0e1a04eba46690e3ffbf5a6e06051 but since it has not been exported, nor used anywhere in Base, I guess it is safe to remove.

The functionality though, i.e., static partitioning of the work load across workers can be achieved in the new `pmap` via batching, right? Maybe provide a keyword arg `partition=fixed` to expose it?

amitmurthy commented on base/pmap.jl:
```diff
@@ -0,0 +1,97 @@
+For multiple collection arguments, apply f elementwise.
+"""
+pmap(p::WorkerPool, f, c...) = collect(pgenerate(p, f, c...))
+pmap(f, c...) = pmap(default_worker_pool(), f, c...)
```
Actually this line should be removed, no? Just the next definition with `err_retry=true, err_stop=false` should maintain existing behavior.

samoconnor replied:
Yeah, it's deliberately overwritten. The intention is that when the version below with all the `depwarn`s is removed then the simple version remains. Happy to remove it if it that approach doesn't make sense to you.

samoconnor replied:
I just noticed that this is causing a compile warning. I've removed it.

samoconnor commented:
The rename of `split` and unexporting is now done.

amitmurthy commented on base/pmap.jl:
```diff
@@ -0,0 +1,102 @@
+    head, tail = head_and_tail(batches, min_batch_count)
+
+    # If there are not enough batches, use a smaller batch size...
+    if length(head) < min_batch_count
```
`length(batches)` ?

samoconnor replied:
`batches` could be infinite, or it could be expensive or slow to compute the length of `batches`. e.g. If `c` is a stream of database query results coming form a remote server. The point of `head_and_tail` is to collect enough items form `c` to know what batch size to use.

amitmurthy commented on base/pmap.jl:
```diff
@@ -0,0 +1,102 @@
+
+    # If there are not enough batches, use a smaller batch size...
+    if length(head) < min_batch_count
+        head = vcat(head...)
```
I don't follow this...

samoconnor replied:
If `c` is list, then `batches` is list of lists and `head` is list of lists.
`vcat(head...)` turns head back into just a list, so that it can be split up again by `partition(head, batch_size)` at the new batch size.

amitmurthy replied:
Understood. Thanks.

samoconnor replied:
```julia
julia> collect(Base.batchsplit(take(countfrom(1), 20), min_batch_count=5))
5-element Array{Array{Int64,1},1}:
 [1,2,3,4]
 [5,6,7,8]
 [9,10,11,12]
 [13,14,15,16]
 [17,18,19,20]

julia> collect(Base.batchsplit(take(countfrom(1), 20), min_batch_count=1))
1-element Array{Any,1}:
 [1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20]

julia> collect(Base.batchsplit(take(countfrom(1), 20), min_batch_count=2))
2-element Array{Array{Int64,1},1}:
 [1,2,3,4,5,6,7,8,9,10]
 [11,12,13,14,15,16,17,18,19,20]

julia> collect(Base.batchsplit(take(countfrom(1), 20), min_batch_count=50))
20-element Array{Array{Int64,1},1}:
 [1]
 [2]
 [3]
 [4]
 [5]
 [6]
 [7]
 [8]
 [9]
 [10]
 [11]
 [12]
 [13]
 [14]
 [15]
 [16]
 [17]
 [18]
 [19]
 [20]
```

amitmurthy commented on base/pmap.jl:
```diff
@@ -0,0 +1,102 @@
+pmap(p::WorkerPool, f, c...) = collect(pgenerate(p, f, c...))
+#pmap(f, c...) = pmap(default_worker_pool(), f, c...)
+
+function pmap(f, c...; err_retry=nothing, err_stop=nothing, pids=nothing)
```
defaults to match existing values?

samoconnor replied:
On one hand, it is arguable that this should be a breaking change because:

 - both the defaults `error_retry=true` and `error_stop=false` are inconsistent with `map`.
 - `err_retry=true` by default is dangerous if `f` has side-effects.
 - most code that uses pmap won't have a huge number of call-sites (so its quick to fix)
 - the "breakage" is limited to error handling, so the risk is low that broken code will persist.

On the other hand, it might make sense to start with a non-breaking change and make the breaking change in a seperate PR, or defer the breakage until later something like this...

```julia
if VERSION < v"0.5.1"

function pmap(f, c...; err_retry=true, err_stop=false, pids=default_worker_pool())

    if err_retry == true
        depwarn("`err_retry=true` is deprecated, " *
                "use `pmap(retry(f), c; err_retry = false)`.", :pmap)
        f = retry(f)
    end

    if err_stop == false
        depwarn("`err_stop=true` is deprecated, " *
                "use `pmap(@catch(f), c; err_stop = true)`.", :pmap)
        f = @catch(f)
    end

    if pids != default_worker_pool()
        depwarn("`pids` is deprecated, use `pmap(::WorkerPool, f, c).", :pmap)
        p = WorkerPool(pids)
    end

    return pmap(p, f, c...)
end

elseif VERSION < v"0.5.2"

function pmap(f, c...; err_retry=nothing, err_stop=nothing)

    if err_retry != nothing || err_stop != nothing
        depwarn("`err_retry` and `err_stop` are deprecated, " *
                "use `pmap(retry(f), c)` or `pmap(@catch(f, c))`.", :pmap)
    end

    return pmap(default_worker_pool(), f, c...)
end

else # VERSION >= "v0.5.2"

pmap(f, c...) = pmap(default_worker_pool(), f, c...)

end
```

samoconnor replied:
Question: is there any precedence for having a depwarn message that can be suppressed by e.g. a global variable.

e.g.
```julia
    if !haskey(ENV, "JULIA_SUPPRESS_PMAP_WARNING")
        depwarn("`err_retry` and `err_stop` no longer have any effect" *
                "use `pmap(retry(f), c)` or `pmap(@catch(f, c))`." *
                "The default retry and error handling behaviour has changed. See ? pmap." *
                "set ENV["JULIA_SUPPRESS_PMAP_WARNING"] to suppress this warning".)
    end
```

amitmurthy replied:
OK. I guess it makes sense to make it consistent with `map`. And `@catch`/`retry` can be used with `map` / `asyncmap` too.

amitmurthy commented on doc/stdlib/parallel.rst:
```diff
@@ -253,13 +265,23 @@ General Parallel Computing Support
+   For multiple collection arguments, apply f elementwise.
 
-   If ``err_retry`` is ``true``\ , it retries a failed application of ``f`` on a different worker. If ``err_stop`` is ``true``\ , it takes precedence over the value of ``err_retry`` and ``pmap`` stops execution on the first error.
+   Note that ``f`` must be made available to all worker processes; see :ref:`Code Availability and Loading Packages <man-parallel-computing-code-availability>` for details.
```
A mention about `@catch` and `retry` and the functionality they provide (as replacement for `err_retry` and `err_stop`) will help.

samoconnor replied:
How about something like this:
```julia
Note that `err_retry=true` and `err_stop=false` are deprecated,
use `pmap(retry(f), c)` or `pmap(@catch(f), c)` instead
(or to retry on a different worker, use `map(retry(remote(f)), c)`).
```

amitmurthy replied:
OK.

amitmurthy commented on test/abstractarray.jl:
```diff
@@ -398,14 +398,17 @@ function test_map(::Type{TestAbstractArray})
         X = typ[1:10...]
         _typ = typeof(arg_typ(one(typ)))
-        @test map(arg_typ, X) == _typ[1:10...]
+        for mapf in [map, asyncmap, pmap]
```
Probably not a good idea to add `pmap` tests here. `runtests.jl` removes workers once they cross a memory threshold. This was one of the reasons parallel tests are being executed outside of the initial cluster. `parallel.jl` explicitly launches `parallel_exec.jl` in another cluster.

It is safer to additional `pmap` tests to `parallel_exec.jl` and test all conditions including effects of worker removal during a `pmap`. You should also modify `pmap` tests run under `JULIA_TESTFULL=1` and test them locally as part of this PR.

samoconnor replied:
In the latest commit, I've removed use of workers from `test/abstractarray.jl` and updated the `JULIA_TESTFULL=1` tests in `test/parallel_exec.jl`.

amitmurthy commented:
I'll be incommunicado till the 6th of April (out backpacking). @ViralBShah / @JeffBezanson / @vtjnash please chime in here and merge this PR once it is ready.

tkelman commented:
`pmap` tests still need to move, and the `generate` wrapper, while no longer exported, still does not serve much purpose. A different name for `pgenerate` might be a good idea considering the unfortunate similarity to `@generated`

samoconnor commented on base/pmap.jl:
```diff
@@ -93,9 +75,8 @@ function batchsplit(c; min_batch_count=1, max_batch_size=100)
-        return partition(head, batch_size)
+        batch_size = max(1, div(sum(length, head), min_batch_count))
+        return partition(flatten(head), batch_size)
     end
```
@amitmurthy with reference to [our previous `vcat` discussion](https://github.com/JuliaLang/julia/pull/15409/files/b845ea1ec5450003f1f749013ae482faac8d7329#r57449634), I'm now using `flatten` instead of `vcat` in `batchsplit` because `vcat` doesn't do the expected thing with strings. (See also https://github.com/JuliaLang/julia/pull/14383#issuecomment-202596418)

tkelman commented on base/pmap.jl:
```diff
@@ -38,38 +38,20 @@ workers and tasks.
 
-    return pmap(p, f, c...)
+if VERSION >= v"0.5.1"
+    error("FIXME: time to remove pmap from deprecated.jl!")
```
No. Deprecations don't get removed in point releases.

samoconnor commented:
Rebased WRT: https://github.com/JuliaLang/julia/pull/15146

nalimilan commented on base/error.jl:
```diff
@@ -49,3 +49,50 @@ macro assert(ex, msgs...)
+for exceptions of that type. If `condition` is a function
+`cond(::Exception) -> Bool` then retry only if it is true.
+
+e.g. `retry(http_get, e->e.status == "503")(url)` or `retry(read, UVError)(io)`.
```
Make this an `# Examples` section (and ideally, with reproducible code).

nalimilan commented on base/error.jl:
```diff
@@ -49,3 +49,50 @@ macro assert(ex, msgs...)
+Returns a lambda that executes `f` and returns either the result of `f` or
+an `Exception` thrown by `f`.
+
+e.g. `@catch(length)() == MethodError(length,())`
```
Same here. See the new guidelines at http://docs.julialang.org/en/latest/manual/documentation/

tkelman commented on base/pmap.jl:
```diff
@@ -0,0 +1,83 @@
+pmap(p::WorkerPool, f, c...) = collect(pgenerate(p, f, c...))
+
+if VERSION >= v"0.5.1"
+    error("FIXME: time to remove pmap from deprecated.jl!")
```
Do not leave time bombs in the code base please. Removing deprecations is a breaking change which we do not do in point releases.

samoconnor commented:
Rebased again to resolve NEWS.md conflict.

samoconnor commented:
I've rebased and squashed a bunch of intermediate commits. Travis is passing. 64-bit AV is passing.
Amit is away backpacking ( lucky him ;) ).
@ViralBShah, @JeffBezanson ready to merge?

tkelman commented on base/generator.jl:
```diff
@@ -22,6 +22,16 @@ function next(g::Generator, s)
+Return an iterator applying `f` to each element of `c`.
+For multiple collection arguments, apply f elementwise.
+"""
+generate(f, c...) = Generator(f, c...)
```
This definition serves no purpose. As long as this is the only method of `generate`, there's no need for this to have 2 names.

tkelman commented on test/error.jl:
```diff
@@ -0,0 +1,72 @@
+
+
+let
+
```
These excess empty lines at the start and end of `function`, `let`, etc blocks are really visually distracting and serve no purpose. Please conform to the prevailing style used everywhere else in the code base that does not do this.

nalimilan commented on test/abstractarray.jl:
```diff
@@ -389,48 +389,52 @@ Base.iteratorsize{N}(::Type{GenericIterator{N}}) = Base.SizeUnknown()
+        C = Float64[1:10...]
+        @test mapf(f, convert(Vector{Int},B), C) == Float64[ 2 * i for i in 1:10 ]
+        @test mapf(f, Int[], Float64[]) == Union{}[]
+        # map with different result tyoes
```
Typo here.

samoconnor replied:
The typo was there already, I didn't notice (I'm a bit dyslexic), but I'll push a fix for it.

nalimilan commented on test/abstractarray.jl:
```diff
@@ -389,48 +389,52 @@ Base.iteratorsize{N}(::Type{GenericIterator{N}}) = Base.SizeUnknown()
+        @test mapf(f, Int[], Float64[]) == Union{}[]
+        # map with different result tyoes
+        let m = mapf(x->x+1, Number[1, 2.0])
+            # FIXME why is this different for asyncmap?
```
Is this FIXME supposed to be fixed in this PR?

samoconnor replied:
no

nalimilan commented on base/pmap.jl:
```diff
@@ -0,0 +1,79 @@
+"""
+function batchsplit(c; min_batch_count=1, max_batch_size=100)
+
+    # FIXME Use @require per https://github.com/JuliaLang/julia/pull/15495
```
If `@require` is added at some point, most of the code base will have to be ported, so there's no point in leaving comments like this. Here, you should raise `ArgumentError`s instead of calling `@assert`.

nalimilan commented on base/multi.jl:
```diff
@@ -1514,101 +1514,10 @@ macro everywhere(ex)
-    Any[ remotecall(f, PGRP.workers[(i-1)%np+1].id, map(L->L[i], lsts)...) for i = 1:n ]
-end
-
+#FIXME delete?
```
?

samoconnor replied:
The "FIXME delete?" refers to `pmap(f) = f()` and referenced commit comment below, do you think I should just delete it?

The pmap_static() was removed following discussion with Amit.

nalimilan replied:
I'd remove the comment, and file an issue if you think this needs to be discussed/fixed.

samoconnor replied:
done #15710

nalimilan commented on base/iterator.jl:
```diff
@@ -447,3 +466,59 @@ end
+    n::Int
+end
+
+eltype{T<:Union{Vector,AbstractString}}(::Type{PartitionIterator{T}}) = T
```
Didn't we say it didn't work for strings currently? So better not accept them for now.

Also, it seems to me that for `Vector` this method is equivalent to the general one below, right?

nalimilan replied:
There's also an example with a string above.

samoconnor replied:
True, and yes. I'll push a change accordingly.

nalimilan commented on NEWS.md:
```diff
@@ -54,6 +54,12 @@ Breaking changes
     is now divided among the fields `code`, `slotnames`, `slottypes`, `slotflags`,
     `gensymtypes`, `rettype`, `nargs`, and `isva` in the `LambdaInfo` type ([#15609]).
 
+  * `pmap` keyword arguments `err_retry=true` and `err_stop=false` are depricated.
```
"deprecated"

samoconnor replied:
thx

nalimilan commented on NEWS.md:
```diff
@@ -54,6 +54,12 @@ Breaking changes
     `gensymtypes`, `rettype`, `nargs`, and `isva` in the `LambdaInfo` type ([#15609]).
 
+  * `pmap` keyword arguments `err_retry=true` and `err_stop=false` are depricated.
+    `pmap` no longer retries or returns `Exception` objects in the result collection.
```
"retries nor returns"?

samoconnor replied:
No strong opinion on this. My feeling is that if it was "neither retries", then it would be "nor returns". I believe that in the "does not verb or verb" case, there is no hard rule about "or" or "nor". When I read it out aloud as written, it sounds better to me with "or" rather than "nor".  However, I'm Australian, so what sounds normal to me would make an Englishman cringe! Happy to change it if you feel strongly ...

nalimilan commented on base/asyncmap.jl:
```diff
@@ -145,4 +145,27 @@ function next(itr::StreamMapIterator, state::StreamMapState)
+
+Results are returned in order as they become available.
+"""
+asyncgenerate(f, c...) = StreamGenerator(f, c...)
```
I must say I find it quite confusing that `asyncgenerate` creates a `StreamGenerator` and not an `AsyncGenerator`. Shouldn't `AsyncGenerator` be `AsyncGenerator!`, to parallel `map`/`map!`? Then `StreamGenerator` could be `AsyncGenerator`.

I don't understand why the concept of "stream" is used here: AFAIK we only use this term for `IOStream`, which doesn't seem to be related.

samoconnor replied:
I see what you mean about `map/map!`.
I worry a little about having a type name with a bang "!" a the end. Is there any precedent for that?

As to why it is the way it is, the names `AsyncMapIterator` and `StreamMapIterator` pre-date the `Generator` name.
The former applies the mapped function with `@async`. The latter makes the results available as a stream (i.e. as they become available, rather than having to wait until the end).

I think I agree that the thing that is currently called `StreamGenerator` should probably be called `AsyncGenerator`.
I could roll these together so that the type currently called `AsyncGenerator` goes away? what do you think?

nalimilan replied:
> I think I agree that the thing that is currently called StreamGenerator should probably be called AsyncGenerator.
> I could roll these together so that the type currently called AsyncGenerator goes away? what do you think?

How would that work?

samoconnor replied:
The thing that uses `AsyncGenerator` is `StreamGenerator`. So, the code in the `AsyncGenerator` type and methods could be rolled into the `StreamGenerator` type and methods and there would be no externally visible difference.

nalimilan replied:
I really can't tell whether that's OK or not, maybe @amitmurthy?

tkelman replied:
If the primary difference is whether the results are saved in place vs not, then the non in place version might not need its own separate wrapper type. It could be convenient to provide both in place and non in place API options but that would more typically be done as a lower case `!` function rather than a type constructor.

As Stefan said in the other issue, something still feels off about the proliferation of names for slight variations here.

samoconnor replied:
@nalimilan I've attempted to resolved the naming confusion with minimal code change in the latest commit.

nalimilan replied:
Thanks, but how about your original @tkelman's suggestion of having only one type?

samoconnor replied:
I started along those lines, but ended up concluding that the implementation just becomes needlessly complicated if the two layers are merged. i.e. it makes sense to have two distinct architectural layers in the implementation. One runs stuff asynchronously and collects the results in whatever order they happen to be computed, the other manages in-order iteration over the results collection. I am open to suggestions on naming. However, these names are not exported, and the basic design was reviewed and accepted already in #15058. I suggest that these names are good enough.

[This line](https://github.com/JuliaLang/julia/pull/15409/commits/c6bc3680d15d9b5762d6c7a1fad62d765247285e#diff-9d7ce9b013c358a3f6e8126bbdc03145R179) demonstrates how the `AsyncCollector` is useful on its own.

I agree that the API could be made simpler, I've commented on that over here: https://github.com/JuliaLang/julia/issues/14843#issuecomment-202656425 
But, I think in this instance the implementation is simpler with `AsyncCollector` kept seperate from `AsyncGenerator`.

nalimilan commented on base/pmap.jl:
```diff
@@ -0,0 +1,79 @@
+
+For multiple collection arguments, apply f elementwise.
+
+Note that `err_retry=true` and `err_stop=false` are deprecated,
```
I don't think documenting deprecated arguments is a good idea. They'll likely remain in the docs even after support has been removed.

samoconnor replied:
This note was added at Amit's request: https://github.com/JuliaLang/julia/pull/15409#discussion_r57496110

nalimilan replied:
Well, yeah. :-) Maybe to please us both, you could mention only the new `retry` and `@catch`, and not the deprecated arguments.

samoconnor replied:
If you care to write some copy that you think would keep everyone happy, I'll gladly paste it in :)

nalimilan replied:
I don't really know how all of this works, but here's a rough idea:
```
Use `pmap(retry(f), c)` to retry a failed application of ``f`` on the same worker,
and `asyncmap(retry(remote(f)), c)`) to retry on a different worker.
Use `pmap(@catch(f), c)` if you want any thrown exception to be returned
as a result.
```

nalimilan commented on base/pmap.jl:
```diff
@@ -0,0 +1,79 @@
+for details.
+"""
+pmap(p::WorkerPool, f, c...) = collect(pgenerate(p, f, c...))
+#pmap(f, c...) = pmap(default_worker_pool(), f, c...)  -- See depricated.jl
```
Same: when we will remove the code from deprecated.jl, we won't notice this line.

samoconnor replied:
I agree. It was previously written as:
```julia
if VERSION >= v"0.5.1"
    error("FIXME: time to remove pmap from deprecated.jl!")
    pmap(f, c...) = pmap(default_worker_pool(), f, c...)
```
... but was told "Do not leave time bombs in the code base", so I removed it.

On the other hand, as it stands, when `pmap` is removed from deprecated, `pmap` will stop working completely. Which will be noticed fairly quickly, so it is not too bad as is.

nalimilan replied:
> On the other hand, as it stands, when pmap is removed from deprecated, pmap will stop working completely. Which will be noticed fairly quickly, so it is not too bad as is.

I don't understand. If this definition is in this file and not in deprecated.jl, it's because it should remain supported, right?

samoconnor replied:
It is the keyword arguments that are deprecated. But as far as I know, Julia's multiple dispatch system does not distinguish between a method with or without keyword arguments, so this line as to be commented out until the `pmap` in deprecated.jl is removed. Both definitions do the same thing aside from the kw args.

nalimilan replied:
OK, keep it that way then, but maybe make the comment a bit more explicit, and explain the action to take in deprecated.jl too.

nalimilan replied:
Still not fixed.

tkelman replied:
also typo in comment, should be deprecated.jl not depricated.jl

tkelman commented on base/asyncmap.jl:
```diff
@@ -0,0 +1,187 @@
+
+Note: next(::AsyncCollector, state) -> (nothing, state)
+
+Note: `for task in AsyncCollector(f, results, c...)` end is equivalent to
```
backtick should include the `end` - also `next(::AsyncCollector, state)` above should be backtick-highlighted

samoconnor replied:
well spotted. fix pushed.

nalimilan commented on test/functional.jl:
```diff
@@ -239,3 +239,33 @@ end
+@test_throws AssertionError Base.partition([1,2,3,4,5], 0)
+@test_throws AssertionError Base.partition([1,2,3,4,5], -1)
+
+@test join(map(x->string(x...), Base.partition("Hello World!", 5)), "|") ==
```
This one should likely go away too.

samoconnor replied:
This seems to work fine now that it is not using character indexing (it is using character iteration now).
See this example where `partition` does the expected thing with the UTF-8 emoji, but string indexing causes an error:

`foo.jl`:
```julia
s = "Monkey 🙈🙊🙊"
@show s
v = collect(Base.partition(s, 4))
@show v
v = collect(Base.partition(s, 3))
@show v
v = collect(Base.partition(s, 2))
@show v
v = collect(Base.partition(s, 1))
@show v

v = collect(Base.partition(s, 8))
@show v

@show s[8]
@show s[12]
@show try s[9] catch e e end
```

```bash
$ ./julia foo.jl
s = "Monkey 🙈🙊🙊"
v = [['M','o','n','k'],['e','y',' ','🙈'],['🙊','🙊']]
v = [['M','o','n'],['k','e','y'],[' ','🙈','🙊'],['🙊']]
v = [['M','o'],['n','k'],['e','y'],[' ','🙈'],['🙊','🙊']]
v = [['M'],['o'],['n'],['k'],['e'],['y'],[' '],['🙈'],['🙊'],['🙊']]
v = [['M','o','n','k','e','y',' ','🙈'],['🙊','🙊']]
s[8] = '🙈'
s[12] = '🙊'
try s[9] catch e e end = UnicodeError: invalid character index
```

nalimilan replied:
OK, great! But then please add non-ASCII characters to the string so that any regression would be caught.

samoconnor replied:
done

amitmurthy commented:
Looks good. The few pending things that I can see:

- merge conflicts
- AV test to be run again
- Tony's concerns on `generate` and `asyncgenerate` - I agree with him on this
- remove comments regarding `@require` - generally not a good idea to mention a PR that may or may not be merged.

samoconnor commented:
Welcome back @amitmurthy I hope you enjoyed the backpacking!

>Looks good. The few pending things that I can see:
>merge conflicts

done

>AV test to be run again

fingers crossed... (perhaps this [MS Windows Subsystem for Linux thing](https://blogs.windows.com/buildingapps/2016/03/30/run-bash-on-ubuntu-on-windows/) could be leveraged in some way to remove the need for Julia Base to support Win32).

>Tony's concerns on generate and asyncgenerate - I agree with him on this

>> Tony wrote: [`generate(f, c...) = Generator(f, c...)`] serves no purpose. As long as this is the only method of generate, there's no need for this to have 2 names. Same with asyncgenerate.

This was included for consistency. There seems to be a convention that functions returning an iterator object have a lower-case name.

```julia
enumerate(itr) = Enumerate(itr)
zip(a, b, c...) = Zip(a, zip(b, c...))
filter(flt, itr) = Filter(flt, itr)
rest(itr,state) = Rest(itr,state)
take(xs, n::Int) = Take(xs, n)
drop(xs, n::Int) = Drop(xs, n)
cycle(xs) = Cycle(xs)
product(a, b, c...) = Prod(a, product(b, c...))
flatten(itr) = Flatten(itr)
eachline(stream::IO) = EachLine(stream)
```

 This convention makes sense to me because it separates interface from implementation. `enumerate` and `zip` are public functional interfaces. `Enumerate` and `Zip` (and `Zip1` and `Zip2`) are private implementation types. The interface doc for these functions say things like "An iterator that ...", or "returns an iterable....". The uses does not need to know about the type of the iterator, they only need to know what functionality it provides.

>remove comments regarding @require - generally not a good idea to mention a PR that may or may not be merged.

done

tkelman commented on base/pmap.jl:
```diff
@@ -0,0 +1,78 @@
+function batchsplit(c; min_batch_count=1, max_batch_size=100)
+
+    @assert min_batch_count > 0
+    @assert max_batch_size > 1
```
throw an ArgumentError

tkelman commented:
That convention does not need to be followed when the lower-case versions are not being exported as public API's.

amitmurthy commented:
Thanks. Could you squash the commits? Will merge in a day or two if there are no major concerns.

samoconnor commented:
squashed

nalimilan commented on base/asyncmap.jl:
```diff
@@ -0,0 +1,187 @@
+end
+
+function next(itr::AsyncGenerator, state::AsyncGeneratorState)
+
```
I don't think a line break is in order here, even if we allowed the exception of blocks starting with a comment.

nalimilan commented on base/deprecated.jl:
```diff
@@ -1003,5 +1003,33 @@ export call
+
+#15409
+function pmap(f, c...; err_retry=nothing, err_stop=nothing, pids=nothing)
+
```
Same here (and likely elsewhere).

nalimilan commented on base/deprecated.jl:
```diff
@@ -1003,5 +1003,33 @@ export call
+    end
+
+    if err_stop != nothing
+        depwarn("`err_stop` is deprecated, use `pmap(@catch(f), c...).", :pmap)
```
Missing closing backquote. BTW, are you sure backquotes in error messages are the usual way of printing exceptions? I think I made a PR about that ages ago which was rejected.

samoconnor replied:
I've removed the backquotes.

samoconnor committed:
cosmetics

amitmurthy commented:
Thanks @samoconnor for this, and everyone for reviewing it. I'll submit a PR for the remaining cosmetic changes discussed here.

samoconnor commented:
Thanks Amit

JeffBezanson commented:
Looking at #15661 made me notice the `@catch` macro. We might not want macros with the same names as keywords. Happily, I also notice it doesn't really need to be a macro, since it just expands to a function call on the same argument. I think this functional style, where possible, is far preferable to macros. Can we just have the function and not the macro? I would maybe call it `caught(f)`.

samoconnor commented:
@JeffBezanson I though that someone might notice that `@catch` is not really a macro.
I was a little surprised that none of the reviewers mentioned it given that the review was thorough enough to get all the way down to discussing positioning of whitespace :)

I chose `@catch` because:
 - I really wanted `catch(f) = ...`, but `catch` is a keyword.
 - I thought that the `@` doesn't hurt, because it conveys a bit of "something not quite normal might happen here". i.e. the return value might be an ordinary value, or it might be an Exception.
 - I couldn't think of an alternative name that retained a strong association with the metaphor of throwing and catching.

I'm not sure that `caught` is quite right. The thing that is caught is the exception. The function does not return an exception, it returns a function. The returned function does not always return an exception either, it mostly returns an ordinary value. The name `catch` seems logical because the returned function does something close to what a catch block does (sometimes nothing, sometimes catching).

My preference would be to remove `@catch` completely.
The only reason for including it was as a way to extract the old exception-as-value feature from `pmap`. e.g. `pmap(f, c, error_stop = false)` is replaced by `pmap(@catch(f), c)`.
However, the user could just be asked to write `pmap(x->try f(c) catch ex; ex end, c)` instead.

Also, why not make `try f(x) catch end` return the exception? (it currently returns `Void`).

Then you would just have `pmap(x->try f(x) catch end, c)`.

Note: Midori has [`v = try Foo() else catch`](http://joeduffyblog.com/2016/02/07/the-error-model/#syntactic-sugar).

nalimilan commented:
You could call it `trycatch`, but I agree we could as well do without it.

tkelman commented:
Could also call it `except(f)` if it needs to have a name. Making try-catch return the exception by default feels like a pretty arbitrary thing to do.

amitmurthy commented:
+1 for `except(f)`

nalimilan commented:
What is "except" supposed to mean? "exception" or "except"?

JeffBezanson commented:
Taking this in a slightly different direction, it makes me nervous to have an API that returns either a value or an exception that was thrown, without leaving any indication of which it was.

tkelman commented:
Something more Nullable-like, a la ExceptionOr would be easier to make type stable and less ambiguous if the non error branch happens to want to return an exception object.

amitmurthy commented:
This is useful in situations where the programmer is OK with a certain amount of errors. For example, a search engine, implementing a distributed map-reduce may be OK displaying the results (instead of an error) if the number of failed calls were less than a certain threshold. Or you would want to a really long running `pmap` to complete and just log any exceptions. Again, you may throw an error only if the number of exceptions crossed a certain threshold.

As another example, there is an open issue for always completing the entire CI test suite even when we encounter a failure.

This is a convenience wrapper for a functionality that users can always code into the mapping function.

tkelman commented:
These cases probably shouldn't be returning exceptions inline among the results though. Maybe a NullableArray with an auxiliary Any-array to hold any exceptions that occur would have more predictable behavior.

amitmurthy commented:
The exceptions array will have to be the same length as the NullableArray - in order to maintain the same indexes as the jobs array and easily track failed jobs - or we return an array of tuples `(idx, excep)`. How would one enable this since It is preferable to have the default behavior same as `map`? I am OK with bringing back a keyword arg for this functionality, say `catch_exceptions=true`, which would enable this behavior. Default is `false`.

 Only if `catch_exceptions=true`, we could either return exceptions inline or return a tuple `(results, exceptions)` where results is a Nullable Array and exceptions an array of `(idx, excep)`.

JeffBezanson commented:
Yes, while you can test `isa(x, Exception)`, this does not actually determine whether an exception happened. However I suppose we could require this by convention.

samoconnor commented:
I don't think there is a compelling reason to have re-introduce `catch_exceptions=true` to `pmap`.

If doing `pmap(x -> try f(x) catch ex; ex end, c)` and `isa(x, Exception)` is not enough for some use-case, then something like this could be done (and be useful in places other than `pmap`):

```julia
witherr(f) = (args...) -> try (f(args...), nothing) catch ex; (nothing, ex) end

julia> v, err = witherr(string)(7)
("7",nothing)

julia> v, err = witherr(error)(7)
(nothing,ErrorException("7"))

pmap(witherr(f), c)
```

amitmurthy commented:
The scenarios which I described above requires `pmap` to be robust against network errors or workers dying too. I'll put a PR together and we can continue discussion on the interface there.

amitmurthy commented:
See 15975
